x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/CartesianPoint.java

```diff
@@ -0,0 +1,239 @@
+            throw new ElasticsearchParseException("[geo_point] supports only POINT among WKT primitives, " +
+                "but found " + geometry.type());
+        }
+        org.elasticsearch.geometry.Point point = (org.elasticsearch.geometry.Point) geometry;
+        return reset((float) point.getX(), (float) point.getY());
+    }
+
+    public float getX() {
+        return this.x;
+    }
+
+    public float getY() {
+        return this.y;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        CartesianPoint point = (CartesianPoint) o;
+
+        if (Float.compare(point.x, x) != 0) return false;
+        if (Float.compare(point.y, y) != 0) return false;
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result;
+        int temp;
+        temp = x != +0.0f ? Float.floatToIntBits(x) : 0;
+        result = Integer.hashCode(temp);
+        temp = y != +0.0f ? Float.floatToIntBits(y) : 0;
+        result = 31 * result + Integer.hashCode(temp);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return x + ", " + y;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder.startObject().field("x", x).field("y", y).endObject();
+    }
+
+    public static CartesianPoint parsePoint(XContentParser parser, CartesianPoint point)
+        throws IOException, ElasticsearchParseException {
+        float x = Float.NaN;
+        float y = Float.NaN;
+        NumberFormatException numberFormatException = null;
+
+        if(parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            try (XContentSubParser subParser = new XContentSubParser(parser)) {
+                while (subParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                    if (subParser.currentToken() == XContentParser.Token.FIELD_NAME) {
+                        String field = subParser.currentName();
+                        if ("x".equals(field)) {
+                            subParser.nextToken();
+                            switch (subParser.currentToken()) {
+                                case VALUE_NUMBER:
+                                case VALUE_STRING:
+                                    try {
+                                        x = subParser.floatValue(true);
+                                    } catch (NumberFormatException e) {
+                                        numberFormatException = e;
+                                    }
+                                    break;
+                                default:
+                                    throw new ElasticsearchParseException("x must be a number");
+                            }
+                        } else if ("y".equals(field)) {
```

Comment:
We should probably make `x` and `y` a static `ParseField` type that can be reused?